// adamab/src/ai.js
import { profile, projects, experience, links } from './data.js';

function buildPortfolioContext() {
  const about = Array.isArray(profile.about) ? profile.about.filter(Boolean).join(' ') : '';
  const projectLines = projects
    .map((p) => `- ${p.name}: ${p.desc} (${p.tech}). Link: ${p.link}`)
    .join('\n');
  const expLines = experience
    .map((e) => `- ${e.role} @ ${e.company} (${e.location}, ${e.period}): ${e.desc}`)
    .join('\n');

  return [
    `Name: ${profile.name}`,
    `Tagline: ${profile.tagline}`,
    about ? `About: ${about}` : '',
    '',
    'Projects:',
    projectLines,
    '',
    'Experience:',
    expLines,
    '',
    'Links:',
    `- GitHub: ${links.github}`,
    `- LinkedIn: ${links.linkedin}`,
    `- Twitter: ${links.twitter}`,
    `- Email: ${links.email}`,
  ]
    .filter(Boolean)
    .join('\n');
}

function toOpenAIMessages(chatMessages) {
  // We only forward user/assistant conversational turns.
  return chatMessages
    .filter((m) => {
      const role = (m?.role || '').toLowerCase();
      return role === 'user' || role === 'assistant';
    })
    .map((m) => ({
      role: m.role.toLowerCase(),
      content: String(m.content || ''),
    }));
}

function getProxyUrl() {
  return (
    process.env.AI_ADAM_PROXY_URL ||
    process.env.PORTFOLIO_PROXY_URL ||
    'https://www.adam-ab.com/api/ai-adam/chat'
  );
}

async function callProxy({ chatMessages, userMessage }) {
  const url = getProxyUrl();
  const controller = new AbortController();
  const timeoutMs = Number(process.env.AI_ADAM_PROXY_TIMEOUT_MS || 15000);
  const timeout = setTimeout(() => controller.abort(), timeoutMs);
  let resp;
  try {
    resp = await fetch(url, {
      method: 'POST',
      signal: controller.signal,
      headers: {
        'content-type': 'application/json',
      },
      body: JSON.stringify({
        message: userMessage,
        history: toOpenAIMessages(chatMessages),
        portfolioContext: buildPortfolioContext(),
      }),
    });
  } finally {
    clearTimeout(timeout);
  }

  const data = await resp.json().catch(() => ({}));
  if (!resp.ok) {
    return {
      ok: false,
      text: data?.error || `Proxy error (${resp.status})`,
    };
  }
  const text = String(data?.text || '').trim();
  return text ? { ok: true, text } : { ok: false, text: 'Empty response from proxy.' };
}

export async function generateAssistantReply({ chatMessages, userMessage }) {
  try {
    return await callProxy({ chatMessages, userMessage });
  } catch (e) {
    return {
      ok: false,
      text:
        `AI backend unavailable: ${e?.message || String(e)}. ` +
        'Check AI_ADAM_PROXY_URL and ensure the proxy is deployed.',
    };
  }
}
