// adamab/src/cli.js
import readline from 'node:readline';
import open from 'open';
import {
  menuPage,
  projectsPage,
  experiencePage,
  chatPage,
  linksPage,
  aboutPage,
  themePage,
  goodbyePage,
} from './ui.js';
import { setTheme, getThemeIds } from './themes.js';
import { projects, experience, links } from './data.js';
import { generateAssistantReply } from './ai.js';

const menuTargets = ['chat', 'experience', 'projects', 'links', 'about', 'theme', 'exit'];

const linkUrls = [
  links.github,
  links.linkedin,
  links.twitter,
  `mailto:${links.email}`,
];

const pageLengths = {
  menu: menuTargets.length,
  projects: projects.length,
  experience: experience.length,
  chat: 0,
  links: linkUrls.length,
  about: 0,
  theme: getThemeIds().length,
};

const renderers = {
  menu: menuPage,
  projects: projectsPage,
  experience: experiencePage,
  chat: chatPage,
  links: linksPage,
  about: aboutPage,
  theme: themePage,
};

const ALT_ON = '\x1B[?1049h';
const ALT_OFF = '\x1B[?1049l';
const HIDE = '\x1B[?25l';
const SHOW = '\x1B[?25h';
const HOME = '\x1B[H';
const WRAP_OFF = '\x1B[?7l';
const WRAP_ON = '\x1B[?7h';

export async function run() {
  let page = 'menu';
  let idx = 0;
  let scrollOffset = 0;
  let chatInput = '';
  let chatScroll = 0;
  let chatCmdIdx = 0;
  let chatCmdScroll = 0;
  let chatPaletteOpen = false;
  let chatPaletteMode = 'commands'; // 'commands' | 'prompts'
  let chatTabPrefix = null;
  let chatTabMatches = [];
  let chatTabIndex = 0;
  let chatSpinner = null;
  let chatBusy = false;
  const chatMessages = [
    {
      role: 'assistant',
      content:
        "Hey — I’m Adam (well, a fake terminal version). Ask me about my projects, experience, hobbies, or anything you’re curious about.",
    },
  ];

  const write = (s) => process.stdout.write(s);

  write(ALT_ON + HIDE + WRAP_OFF);

  readline.emitKeypressEvents(process.stdin);
  if (process.stdin.isTTY) process.stdin.setRawMode(true);
  process.stdin.resume();

  const draw = () => {
    let out;
    let cursor = null;
    if (page === 'projects' || page === 'experience') {
      out = renderers[page](idx, scrollOffset);
      if (out && typeof out === 'object' && 'frame' in out) {
        scrollOffset = out.scrollOffset;
        out = out.frame;
      } else {
        out = String(out);
      }
    } else if (page === 'chat') {
      const commands = [
        { command: '/quit', description: 'exit chat' },
        { command: '/back', description: 'back to menu' },
        { command: '/clear', description: 'clear transcript' },
        { command: '/help', description: 'show commands' },
        { command: '/prompts', description: 'premade prompts' },
      ];
      const premadePrompts = [
        { title: 'Quick intro', text: 'Give me a quick intro about you in 3 bullets.' },
        { title: 'Projects overview', text: 'What projects are you most proud of, and why?' },
        { title: 'Experience highlights', text: 'Summarize your experience and biggest impact.' },
        { title: 'Tech stack', text: 'What technologies do you like working with and why?' },
        { title: 'Hobbies', text: 'What are your hobbies and what do you enjoy about them?' },
        { title: 'Contact', text: 'What’s the best way to reach you?' },
      ];
      const query = chatInput.trim();
      const paletteQuery = query.startsWith('/') ? query : '';
      const promptsQuery = (() => {
        if (!paletteQuery.startsWith('/prompts')) return '';
        const rest = paletteQuery.replace(/^\/prompts\s*/i, '');
        return rest.trim().toLowerCase();
      })();
      if (paletteQuery.startsWith('/prompts')) chatPaletteMode = 'prompts';

      const commandItems = paletteQuery ? commands.filter((c) => c.command.startsWith(paletteQuery)) : commands;
      const promptItems = premadePrompts
        .filter((p) => {
          if (!promptsQuery) return true;
          return (p.title + ' ' + p.text).toLowerCase().includes(promptsQuery);
        })
        .map((p) => ({ command: p.title, description: p.text, value: p.text }));

      const filtered = chatPaletteMode === 'prompts' ? promptItems : commandItems;
      const palette = chatPaletteOpen
        ? {
            open: true,
            query: chatPaletteMode === 'prompts' ? promptsQuery : paletteQuery,
            items: filtered,
            idx: chatCmdIdx,
            scrollOffset: chatCmdScroll,
            mode: chatPaletteMode,
          }
        : undefined;
      out = renderers.chat({
        messages: chatMessages,
        input: chatInput,
        scrollOffset: chatScroll,
        commandPalette: palette,
      });
      if (out && typeof out === 'object' && 'frame' in out) {
        chatScroll = out.scrollOffset;
        cursor = out.cursor || null;
        out = out.frame;
      } else {
        out = String(out);
      }
    } else {
      out = renderers[page](idx);
    }
    const frame = out.replace(/\n$/, '');
    const lines = frame.split('\n').map((l) => l + '\x1B[K').join('\n');
    if (page === 'chat') {
      write(HOME + lines + '\x1B[0J');
      // Cursor placement is provided by the chat renderer so palette selections don't confuse it.
      if (cursor && cursor.row && cursor.col) {
        write(`\x1B[${cursor.row};${cursor.col}H` + SHOW);
      } else {
        write(SHOW);
      }
    } else {
      write(HIDE + HOME + lines + '\x1B[0J' + SHOW);
    }
  };

  const go = (target) => {
    page = target;
    idx = 0;
    scrollOffset = 0;
    chatInput = '';
    chatScroll = 0;
    chatCmdIdx = 0;
    chatCmdScroll = 0;
    chatPaletteOpen = false;
    draw();
  };

  const cleanup = () => {
    if (process.stdin.isTTY) process.stdin.setRawMode(false);
    process.stdin.pause();
    write(WRAP_ON + ALT_OFF + SHOW);
  };

  const exit = () => {
    cleanup();
    console.log(goodbyePage());
    process.exit(0);
  };

  const back = () => (page === 'menu' ? exit() : go('menu'));

  draw();

  process.stdin.on('keypress', async (ch, key) => {
    const n = key?.name;
    const len = pageLengths[page];

    if (page === 'chat') {
      const commands = [
        { command: '/quit', description: 'exit chat' },
        { command: '/back', description: 'back to menu' },
        { command: '/clear', description: 'clear transcript' },
        { command: '/help', description: 'show commands' },
        { command: '/prompts', description: 'premade prompts' },
      ];
      const premadePrompts = [
        { title: 'Quick intro', text: 'Give me a quick intro about you in 3 bullets.' },
        { title: 'Projects overview', text: 'What projects are you most proud of, and why?' },
        { title: 'Experience highlights', text: 'Summarize your experience and biggest impact.' },
        { title: 'Tech stack', text: 'What technologies do you like working with and why?' },
        { title: 'Hobbies', text: 'What are your hobbies and what do you enjoy about them?' },
        { title: 'Contact', text: 'What’s the best way to reach you?' },
      ];
      const query = chatInput.trim();
      const paletteQuery = query.startsWith('/') ? query : '';
      const promptsQuery = (() => {
        if (!paletteQuery.startsWith('/prompts')) return '';
        const rest = paletteQuery.replace(/^\/prompts\s*/i, '');
        return rest.trim().toLowerCase();
      })();
      if (paletteQuery.startsWith('/prompts')) chatPaletteMode = 'prompts';
      const commandItems = paletteQuery ? commands.filter((c) => c.command.startsWith(paletteQuery)) : commands;
      const promptItems = premadePrompts
        .filter((p) => {
          if (!promptsQuery) return true;
          return (p.title + ' ' + p.text).toLowerCase().includes(promptsQuery);
        })
        .map((p) => ({ command: p.title, description: p.text, value: p.text }));
      const filtered = chatPaletteMode === 'prompts' ? promptItems : commandItems;

      const execCommand = (cmd) => {
        const c = (cmd || '').trim().toLowerCase();
        if (c === '/quit' || c === '/back') return back();
        if (c === '/clear') {
          chatMessages.length = 0;
          chatScroll = 0;
          return draw();
        }
        if (c === '/help') {
          chatMessages.push({
            type: 'status',
            label: 'Commands',
            detail: commands.map((x) => x.command).join('  '),
          });
          chatScroll = 1e9;
          return draw();
        }
        if (c === '/prompts') {
          chatPaletteOpen = true;
          chatPaletteMode = 'prompts';
          chatCmdIdx = 0;
          chatCmdScroll = 0;
          chatInput = '/prompts ';
          return draw();
        }
      };

      const resetTabState = () => {
        chatTabPrefix = null;
        chatTabMatches = [];
        chatTabIndex = 0;
      };

      const normalizePromptsSpacing = () => {
        // Ensure `/prompts` always has a delimiter space before the query.
        // e.g. "/promptsQuic" -> "/prompts Quic"
        chatInput = chatInput.replace(/^\/prompts(\S)/i, '/prompts $1');
      };

      if (n === 'up' && chatPaletteOpen) {
        chatCmdIdx = Math.max(0, chatCmdIdx - 1);
        if (chatCmdIdx < chatCmdScroll) chatCmdScroll = chatCmdIdx;
        resetTabState();
        draw();
        return;
      }
      if (n === 'down' && chatPaletteOpen) {
        chatCmdIdx = Math.min(Math.max(0, filtered.length - 1), chatCmdIdx + 1);
        const maxRows = 8;
        if (chatCmdIdx >= chatCmdScroll + maxRows) chatCmdScroll = Math.max(0, chatCmdIdx - maxRows + 1);
        resetTabState();
        draw();
        return;
      }
      if (chatPaletteOpen && ((key?.ctrl && n === 'u') || n === 'pageup')) {
        const page = 8;
        chatCmdScroll = Math.max(0, chatCmdScroll - page);
        chatCmdIdx = Math.max(0, Math.min(chatCmdIdx, chatCmdScroll + page - 1));
        resetTabState();
        draw();
        return;
      }
      if (chatPaletteOpen && ((key?.ctrl && n === 'd') || n === 'pagedown')) {
        const page = 8;
        const maxScroll = Math.max(0, filtered.length - page);
        chatCmdScroll = Math.max(0, Math.min(maxScroll, chatCmdScroll + page));
        chatCmdIdx = Math.max(chatCmdScroll, Math.min(chatCmdIdx, chatCmdScroll + page - 1));
        resetTabState();
        draw();
        return;
      }
      if (n === 'up') {
        chatScroll = Math.max(0, chatScroll - 1);
        resetTabState();
        draw();
        return;
      }
      if (n === 'down') {
        chatScroll = chatScroll + 1;
        resetTabState();
        draw();
        return;
      }
      if (n === 'tab') {
        if (!paletteQuery) return;

        // Prompts mode: tab cycles the prompt selection and fills `/prompts <title>`
        // so Enter can apply it (or you can keep tabbing).
        if (chatPaletteMode === 'prompts') {
          if (!filtered.length) return;
          chatCmdIdx = (chatCmdIdx + 1) % filtered.length;
          const maxRows = 8;
          if (chatCmdIdx < chatCmdScroll) chatCmdScroll = chatCmdIdx;
          if (chatCmdIdx >= chatCmdScroll + maxRows) chatCmdScroll = Math.max(0, chatCmdIdx - maxRows + 1);
          chatInput = `/prompts ${filtered[chatCmdIdx].command}`;
          normalizePromptsSpacing();
          resetTabState();
          draw();
          return;
        }

        const prefix = paletteQuery.split(/\s+/)[0];
        // Only autocomplete actual slash commands (not prompt titles).
        const options = commandItems.map((c) => c.command);
        const matches = options.filter((c) => c.startsWith(prefix)).sort();
        if (!matches.length) return;

        if (chatTabPrefix !== prefix || chatTabMatches.join('|') !== matches.join('|')) {
          chatTabPrefix = prefix;
          chatTabMatches = matches;
          chatTabIndex = 0;
        } else {
          chatTabIndex = (chatTabIndex + 1) % chatTabMatches.length;
        }
        chatInput = chatTabMatches[chatTabIndex] + ' ';
        chatPaletteOpen = true;
        chatPaletteMode = chatInput.trim().startsWith('/prompts') ? 'prompts' : 'commands';
        chatCmdIdx = 0;
        chatCmdScroll = 0;
        draw();
        return;
      }
      if (n === 'backspace') {
        chatInput = chatInput.slice(0, -1);
        if (!chatInput.startsWith('/')) {
          chatPaletteOpen = false;
          chatPaletteMode = 'commands';
        } else if (!chatInput.trim().startsWith('/prompts')) {
          chatPaletteMode = 'commands';
        }
        normalizePromptsSpacing();
        resetTabState();
        draw();
        return;
      }
      if (n === 'escape') {
        chatPaletteOpen = false;
        chatPaletteMode = 'commands';
        resetTabState();
        draw();
        return;
      }
      if (n === 'return') {
        if (chatBusy) return;
        const msg = chatInput.trim();
        if (msg) {
          if (msg.startsWith('/')) {
            // Execute slash command instead of sending.
            if (chatPaletteOpen && filtered[chatCmdIdx]) {
              if (chatPaletteMode === 'prompts') {
                const val = filtered[chatCmdIdx].value || filtered[chatCmdIdx].description || '';
                chatInput = String(val);
                chatPaletteOpen = false;
                chatPaletteMode = 'commands';
                chatCmdIdx = 0;
                chatCmdScroll = 0;
                draw();
                return;
              }
              execCommand(filtered[chatCmdIdx].command);
            } else {
              execCommand(msg.split(/\s+/)[0]);
            }
          } else {
            chatBusy = true;
            chatMessages.push({ role: 'user', content: msg });
            const thinkingIdx = chatMessages.push({ type: 'thinking', content: 'Thinking' }) - 1;
            chatInput = '';
            chatScroll = 1e9;
            chatPaletteOpen = false;
            chatPaletteMode = 'commands';
            chatCmdIdx = 0;
            chatCmdScroll = 0;
            resetTabState();
            draw();

            const frames = ['Thinking', 'Thinking.', 'Thinking..', 'Thinking...'];
            let frameIdx = 0;
            if (chatSpinner) {
              clearInterval(chatSpinner);
              chatSpinner = null;
            }
            chatSpinner = setInterval(() => {
              if (!chatBusy || !chatMessages[thinkingIdx] || chatMessages[thinkingIdx].type !== 'thinking') {
                clearInterval(chatSpinner);
                chatSpinner = null;
                return;
              }
              chatMessages[thinkingIdx].content = frames[frameIdx];
              frameIdx = (frameIdx + 1) % frames.length;
              chatScroll = 1e9;
              draw();
            }, 250);

            const result = await generateAssistantReply({
              chatMessages,
              userMessage: msg,
            });

            if (chatSpinner) {
              clearInterval(chatSpinner);
              chatSpinner = null;
            }
            chatMessages.splice(thinkingIdx, 1);
            chatMessages.push({
              role: 'assistant',
              content: result.ok ? result.text : result.text,
            });
            chatScroll = 1e9;
            chatBusy = false;
            draw();
            return;
          }
          chatInput = '';
          chatScroll = 1e9;
          chatPaletteOpen = false;
          chatPaletteMode = 'commands';
          chatCmdIdx = 0;
          chatCmdScroll = 0;
          resetTabState();
          draw();
        }
        return;
      }
      if (ch === '/' && !key?.ctrl && !key?.meta) {
        chatPaletteOpen = true;
        chatPaletteMode = 'commands';
        chatCmdIdx = 0;
        chatCmdScroll = 0;
        resetTabState();
      }
      if (ch && typeof ch === 'string' && ch.length === 1 && !key?.ctrl && !key?.meta) {
        chatInput += ch;
        normalizePromptsSpacing();
        if (!chatInput.startsWith('/')) chatPaletteOpen = false;
        if (chatInput.trim().startsWith('/prompts')) {
          chatPaletteOpen = true;
          chatPaletteMode = 'prompts';
        } else if (chatPaletteMode === 'prompts') {
          chatPaletteMode = 'commands';
        }
        resetTabState();
        draw();
        return;
      }
      if (ch === 'q' || ch === 'h' || (key?.ctrl && n === 'c')) {
        back();
        return;
      }
    }

    if ((n === 'up' || ch === 'k') && len > 0) {
      idx = (idx - 1 + len) % len;
      draw();
    } else if ((n === 'down' || ch === 'j') && len > 0) {
      idx = (idx + 1) % len;
      draw();
    } else if (n === 'return' || ch === 'l') {
      if (page === 'menu') {
        const target = menuTargets[idx];
        target === 'exit' ? exit() : go(target);
      } else if (page === 'theme') {
        const themeIds = getThemeIds();
        setTheme(themeIds[idx]);
        go('menu');
      } else if (page === 'projects') {
        await open(projects[idx].link);
      } else if (page === 'links') {
        await open(linkUrls[idx]);
      }
    } else if (page === 'menu' && ch === 'c') {
      go('chat');
    } else if (ch === 'q' || ch === 'h' || (key?.ctrl && n === 'c')) {
      back();
    }
  });
}
